### ISSUE TITLE
extending notebook panel - how to add layout elements?

### ISSUE THREAD

**Issue by rafikam** [jupyterlab/jupyterlab#4467]
Sorry, newbie question here... I was able to use your guide on extending notebook to incorporate a new notebook toolbar button for saving a version to source control. But I also want to give user a UI display of that version history. I'd like to add this as a docked footer or sidebar panel within the notebook panel itself. Is this possible and what's the right way to go about adding this?

From what I can see, layout is handled privately within notebook panel class. Is there a way to add a new widget to it from a lab extension? Thanks

**Comment by jasongrout** [comment]
CC @ian-r-rose, who experimented with adding a side panel when he was working on the table-of-contents extension.

**Comment by ian-r-rose** [comment]
We have talked a few times about having a richer API for adding side panels, footers, etc, to document widgets (see some of the discussion in #3949). Ultimately, we decided to put that on the back-burner a bit while we try to stabilize the core APIs for 1.0.

It may be possible to do something similar to what I did in that (now closed) PR, but it would be a bit of a hack. It involves a bunch of manual DOM manipulation for the notebook widget, which should mostly be considered private, so I would expect it to be pretty brittle.

**Comment by rafikam** [comment]
Thanks for the fast response. Sounds like it will be best for me to approach it by making using the app's left panel until the API is updated. By the way, you may want to add to your extension docs an example of an extension that adds a simple widget to the left panel. It seems like a common use case, so it would be nice to see it in a very bare bones form, the way you've done for the notebook toolbar button example and the xkcd widget.

**Comment by ian-r-rose** [comment]
I agree, it would be nice to have a minimal example that adds something to the left (or right) panel. At the moment, you are right: both the Cell Tools and [jupyterlab-toc](https://github.com/ian-r-rose/jupyterlab-toc) extensions are fairly small examples that do that, and react to the currently active notebook.
